_posts/2018-02-06-unrealiable-network.md

```diff
@@ -0,0 +1,116 @@
+
+**Is some specific request slowing down the initial render of your application?**
+
+Your users want your application to load as quickly as possible, and thus you should only make the network requests that are absolutely mandatory to get your application up and running. If the network infrastructure is unstable, getting the application to load itself can take a long time or even fail completely.
+
+# Retry with Increasing Timeout
+
+Now that we have optimised the network requests, it is time to make sure those requests pass to the server and back as often as possible. Application level network requests always operate on top of lower-level transmission protocols, such as [TCP](https://en.wikipedia.org/wiki/Transmission_Control_Protocol). These protocols try to provide an abstraction of a reliable connection. This is achieved by the sender detecting lost data and retransmitting it to the receiver automatically. So, if lower-level protocols already try to abstract reliable connection in every situation, why would we want to write our own re-try mechanism on application level?
+
+The answer is that application level network requests, such as HTTP requests, can nominally finish, but still transiently get an error code from the other end. This can happen if there are some transient errors in proxies or load balancers. If we cannot fix these problems by making modifications to the network infrastructure, we can abstract more realiable connection on application level.
+
+Rater than calling the application environment's native functions for sending network requests every single time we make a network request, one would consider writing a separate communication API for your application. The idea of this API is to be a wrapper for the environment's native network request functions. The difference is that, with our own API, we can make global modifications on how the requests and responses are handled. For example, we can automatically re-try failing requests.
+
+The idea with re-trying requests is the following: when our communication API is called to send a request to the server, it does this normally and returns the successful response to the caller. However, if the request fails, our API waits a few seconds, and tries to send the request again. If it fails again, it waits even more, and tries again. If the request keeps failing too many times, the API returns the failed request to the caller. In most cases, hopefully, the request is going to be successful after one or two attempts. The user of the API does not know anything about the failed requests. It does not need to, as our API automativally keeps trying sending the requests again. The caller is informed about the failed request only if it has failed multiple times. Here is a naive example of a communication API written in [ClojureScript](https://clojurescript.org):
```

Comment:
> re-trying

retrying

> it waits even more

it waits longer

> If the request keeps failing too many times

If the request fails too many times

> The user of the API does not know anything about the failed requests

The caller of the API does not know of the failed requests.

> automativally keeps trying sending the requests again

automatically keeps (re)sending the requests